JBTM-3911 Remove redundant lock monitor from LRAService

marcosgopen opened the pull request:
https://issues.redhat.com/browse/JBTM-3911

!CORE !AS_TESTS !RTS !JACOCO !XTS !QA_JTA !QA_JTS_OPENJDKORB !PERFORMANCE LRA !DB_TESTS

xstefank commented on rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRAWithParticipantsTest.java:
```diff
@@ -0,0 +1,314 @@
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.Response;
+
+public class LRAWithParticipantsTest {
```
@marcosgopen this is very much like copy-paste of LRATest. Refactor it so we don't need to always change the same thing in two different classes please.

marcosgopen replied:
Thanks Martin, I cleaned and refactored the test.

xstefank commented on rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java:
```diff
@@ -328,7 +328,7 @@ public int leave(URI lraId, String compensatorUrl) {
     }
 
-    public synchronized int joinLRA(StringBuilder recoveryUrl, URI lra, long timeLimit,
+    public int joinLRA(StringBuilder recoveryUrl, URI lra, long timeLimit,
```
As we discussed offline, I'm not sure what all the possible implications of removing synchronized from this method could be. I would like to hear from @mmusgrov why it was synchronized in the first place.

marcosgopen replied:
Good idea, I put the on hold label.

mmusgrov replied:
@marcosgopen I added a comment to the JBTM indicating that the tests are good but I'm not convinced that removing the monitor is the correct approach. Personally I would keep the tests but withdraw the fix and work on a different fix for the lock contention, do you mind if I look into the alternate approach?

marcosgopen commented:
That works for me @mmusgrov , I put the on hold label and I will re-adapt the PR so to keep only the test.

marcosgopen commented on rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/tools/osb/mbean/ObjStoreBrowserLRATest.java:
```diff
@@ -90,7 +90,8 @@ public void lraMBean() throws Exception {
     @Test
     public void lraMBeanRemoval() throws Exception {
-        LongRunningAction lra = new LongRunningAction(new Uid());
+        String lraUrl = "http://localhost:8080/lra";
```
Previous way of creating the LRA was not working anymore, lraID was not initialized and therefore was causing a NullPointerException at https://github.com/jbosstm/narayana/blob/main/rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java#L107. So changing how it is initialized.

mmusgrov replied:
I will take a look next week (I'm on PTO today but wanted to get this PR reviewed asap).

mmusgrov replied:
But don't let this comment block the PR.

mmusgrov replied:
The constructor should only be used when loading an LRA from the ObjectStore. So after constructing the LRA the expectation is that it restore_state will be called so that the URI for the LRA gets initialised. Clearly the code is brittle so it would be nice to firm it up in this area - I'll raise a JBTM so that we don't loose site of it.

marcosgopen replied:
Thanks Mike!

marcosgopen commented on rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:
```diff
@@ -904,20 +904,20 @@ private static String makeLink(String uriPrefix, String key) {
     }
 
-    private void clearObjectStore() {
+    public static void clearObjectStore(TestName testName) {
```
Set this method as static and public so that the new test class can use it too.

tomjenkinson commented on rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:
```diff
@@ -904,20 +904,20 @@ private static String makeLink(String uriPrefix, String key) {
     }
 
-    private void clearDirectory(final File directory) {
+    public static void clearDirectory(final File directory, TestName testName) {
```
+1 to the need for this testName parameter given that the clearDirectory function is essentially being proposed as a utility method for use outside of LRATest class now and because testName is already used in here in the existing implementation of the method.

marcosgopen replied:
I think that is a good idea, I will use a common test base class then

mmusgrov reviewed (changes requested):
The solution looks good but I did have a change request.

marcosgopen commented on rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java:
```diff
@@ -725,24 +725,30 @@ protected LRAStatus toLRAStatus(int atomicActionStatus) {
         }
-
-        return participant;
+        LRALogger.i18nLogger.error_enlistment();
```
> it should also be reported if the internal call to doEnlistParticipant fails, if it isn't already
It is already checked, in fact in that case the participant is null and the caller returns an error: see https://github.com/jbosstm/narayana/blob/main/rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java#L393-L395

mmusgrov replied:
Note also that we used ERROR for reporting the problem back to the client but WARN for logging inside the coordinator since it is an error for the client but not for the coordinator.

marcosgopen replied:
Sorry Mike, I misinterpreted your sentence 'it is an error'. Now I clearly see your reasoning. I will set the level as WARN. Thanks

marcosgopen commented:
@mmusgrov I added two commits, the first for the LRATestBase (common class) and the second for the logMessage annotation (error level). I hope I am addressing all your comments. I am adding this two commits to simplify the review but if required I would then squash them.

mmusgrov commented on rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java:
```diff
@@ -725,24 +725,33 @@ protected LRAStatus toLRAStatus(int atomicActionStatus) {
+            long timeLimit, String compensatorData) throws UnsupportedEncodingException {
+        ReentrantLock lock = tryLockTransaction();
+        if (lock == null) {
+            LRALogger.i18nLogger.error_enlistment();
```
I [said earlier](https://github.com/jbosstm/narayana/pull/2293#discussion_r1750047056) that:

> we used ERROR for reporting the problem back to the client but WARN for logging inside the coordinator since it is an error for the client but not for the coordinator.

and you gave my comment the "thumbs up" but this log message is logged at the coordinator using `@LogMessage(level = ERROR)` instead of `level = WARN`.

marcosgopen replied:
Sorry, my mistake. I'll update the PR.

mmusgrov commented on rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java:
```diff
@@ -725,24 +725,33 @@ protected LRAStatus toLRAStatus(int atomicActionStatus) {
-            // need to remember that there is a new participant
-            deactivate(); // if it fails the superclass will have logged a warning
-            savedIntentionList = true; // need this clean up if the LRA times out
+        else {
```
I don't think any of the following 17 lines of code have changed apart from the indentation has it?

marcosgopen replied:
I can count 14 lines that haven't changed (from line 736 to 749) apart from the indentation. The finally block has been added in this PR.

mmusgrov commented on rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java:
```diff
@@ -725,24 +725,33 @@ protected LRAStatus toLRAStatus(int atomicActionStatus) {
-        if (participant != null) {
-            participant.setCompensatorData(compensatorData);
-            return participant; // must have already been enlisted
+            long timeLimit, String compensatorData) throws UnsupportedEncodingException {
```
Superfluous reformatting.

mmusgrov commented on rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java:
```diff
@@ -725,24 +725,33 @@ protected LRAStatus toLRAStatus(int atomicActionStatus) {
-            return participant; // must have already been enlisted
+            long timeLimit, String compensatorData) throws UnsupportedEncodingException {
+        ReentrantLock lock = tryLockTransaction();
+        if (lock == null) {
```
Thanks

mmusgrov commented on rts/lra/service-base/src/main/java/io/narayana/lra/logging/LraI18nLogger.java:
```diff
@@ -165,6 +165,9 @@ String info_failedToEnlistingLRANotFound(URL lraId, URI coordinatorUri, int coor
     @Message(id = 25039, value = "Invalid argument passed to method: %s")
     String error_invalidArgument(String reason);
 
+    @LogMessage(level = ERROR)
```
The way the message is used implies that it should be logged at `level = WARN`.

marcosgopen replied:
My mistake, sorry. I'll update it.

mmusgrov commented:
@marcosgopen Will you also squash the commits prior to merging please.

marcosgopen commented:
@mmusgrov PR updated and commits squashed. Thanks for your review. 
WDYT if we change the JBTM title with a more consistent one? e.g. "Replace synchronized in favor of Reentrant Lock in LRAService joinLRA method'

mmusgrov reviewed (commented):
PR looks good, sorry for not explaining ERROR vs WARN logging levels more clearly.

mmusgrov commented:
> @mmusgrov PR updated and commits squashed. Thanks for your review. WDYT if we change the JBTM title with a more consistent one? e.g. "Replace synchronized in favor of Reentrant Lock in LRAService joinLRA method'

+1 clearer titles and descriptions help a lot to ease the "cognitive burden".
